BrowserKit/Sources/SiteImageView/SiteImageHandler.swift

```diff
@@ -107,36 +106,42 @@ public class DefaultSiteImageHandler: SiteImageHandler {
         }
     }
 
+    // Note: Must be on an actor to avoid duplicating queued requests.
```

Comment:
@ih-codes I _think_ I understand the intent of this change though I'll probably need to take a deeper dive to grok it fully because I'm not immediately clear on all of the implications. We've changed this class to an `actor` but I'm not really seeing what this gets us since most of the funcs are marked `nonisolated` which is going to remove the state synchronization benefits in those functions AFAIU. Are we using an `actor` here solely to synchronize changes to `requestQueue`? That is confusing to me because I believe that [actors don't protect static state](https://forums.swift.org/t/does-an-actor-protect-static-properties-and-functions/60970/4) since it's effectively global, but maybe that has changed or I'm misunderstanding something.

Reply:
@mattreaganmozilla Here's a fun example of the pattern I was borrowing (might be easier to grok): https://x.com/cocoawithlove/status/1402762956211769347

Basically I only wanted to isolate `getFaviconImage`. My concern was that multiple calls could trigger and I didn't want more than one call to drop past the check to `requestQueue` on line 114. There's no need to protect access elsewhere.

Maybe a better solution would be making a custom actor annotation (something like `@MainActor`) for just that method rather than making the whole class an actor and using a bunch of `nonisolated`. Haven't tried that before.

I'm still a bit new to actors myself if you have any alternative suggestions. Or maybe actors isn't the best way to solve this situation. 🤔

Reply:
My initial thinking was that an `actor` with most funcs marked `nonisolated` is a potential code smell, but I think this is one of the areas of our app architecture where we don't have an official guideline / policy, though I worry that perhaps we're using the wrong tool. 

If we only want to synchronize/lock around `getFaviconImage` there may be better synchronization primitives (an internal lock, queue, or even something like `objc_sync_enter/exit`) we could leverage that would be more clear, as it stands right now I think `actor` is a bit confusing given that there's not really any mutable state on the actor to protect, and `DefaultSiteImageHandler` being Sendable doesn't seem to (AFAICS) currently matter much elsewhere.

Having said all that 😅, I still need to dig in further to make sure I'm fully understanding the change here; I'll defer to you (and Orla) in the meantime. Thanks for investigating all of this, this area of the codebase will benefit greatly from improvements. 🙏

Reply:
Cool, I think I'll go with a synchronization primitive or... I could just annotate the method with `@MainActor`, honestly. 🤔  That's what I had originally. 

I will test this a bit more because I'm not even sure synchronization is a problem to fix in the first place, I was just guessing there was potential a race condition.

Regarding the change as a whole -- the idea is that we keep a queue (or rather, a dictionary) of active favicon requests (which are Tasks to await). Any other code that attempts to get the same favicon will instead await the current Task that's already in the queue. Once the Task completes, all awaiting threads also complete with the same return value from the Task (i.e. the favicon image). Hope that helps clarify a bit! I was testing this in a sample project earlier just for my sanity's sake haha.

Reply:
> Regarding the change as a whole -- the idea is that we keep a queue (or rather, a dictionary) of active favicon requests (which are Tasks to await). Any other code that attempts to get the same favicon will instead await the current Task that's already in the queue

I can't remember the details now, but when I had originally looked into this I recall thinking there was a reason this wouldn't work, because that was what I had originally tried also (but I could have been mistaken). I think the previous hack I implemented was -- effectively -- doing the same thing, except instead of suspending the duplicate task via `return await activeRequest.value`, it was sleeping them (which should also suspend and not block, but it's definitely not a great approach). In any case, if this is working then it's an improvement over my solution so I'm all for it. Thanks for digging into this.